src/GafferArnoldUI/GoboVisualiser.cpp

```diff
@@ -0,0 +1,365 @@
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "IECore/Shader.h"
+#include "IECoreGL/Group.h"
+#include "IECoreGL/Primitive.h"
+#include "IECoreGL/ShaderStateComponent.h"
+#include "IECoreGL/ShaderLoader.h"
+#include "IECoreGL/TextureLoader.h"
+#include "IECoreGL/QuadPrimitive.h"
+#include "IECore/VectorTypedData.h"
+#include "IECore/ObjectVector.h"
+#include "IECoreGL/Renderable.h"
+#include "IECore/LRUCache.h"
+
+#include "Gaffer/Metadata.h"
+#include "GafferSceneUI/StandardLightVisualiser.h"
+#include "GafferSceneUI/LightFilterVisualiser.h"
+#include "GafferOSL/ShadingEngine.h"
+
+#include "boost/format.hpp"
+
+using namespace std;
+using namespace Imath;
+using namespace IECore;
+using namespace IECoreGL;
+using namespace GafferSceneUI;
+
+namespace
+{
+
+CompoundDataPtr evalOSLTexture( IECore::ObjectVectorPtr shaderVector, int resolution );
+
+struct OSLTextureCacheKey
+{
+
+	OSLTextureCacheKey()
+		:	shaders( NULL )
+	{
+	}
+
+	OSLTextureCacheKey( IECore::ObjectVector *s )
+		:	shaders( s )
+	{
+		s->hash( hash );
+	}
+
+	bool operator == ( const OSLTextureCacheKey &other ) const
+	{
+		return hash == other.hash;
+	}
+
+	bool operator != ( const OSLTextureCacheKey &other ) const
+	{
+		return hash != other.hash;
+	}
+
+	bool operator < ( const OSLTextureCacheKey &other ) const
+	{
+		return hash < other.hash;
+	}
+
+	mutable IECore::ObjectVector *shaders;
+	MurmurHash hash;
+
+};
+
+inline size_t tbb_hasher( const OSLTextureCacheKey &cacheKey )
+{
+	return tbb_hasher( cacheKey.hash );
+}
+
+CompoundDataPtr getter( const OSLTextureCacheKey &key, size_t &cost )
+{
+	cost = 1;
+	return evalOSLTexture( key.shaders, 512 );
+}
+
+typedef LRUCache<OSLTextureCacheKey, CompoundDataPtr> OSLTextureCache;
+OSLTextureCache g_oslTextureCache( getter, 10000 );
```

Comment:
What's the memory usage associated with this limit? I make it this :

```
res * yres * channels * sizeof( float ) * cacheSizeLimit
512 * 512 * 3 * 4 * 10000
31457280000 bytes
30 gigabytes
```

Seems a bit much!

I think I'm right in saying that the visualisers are only called again a second time when the attributes have changed? In which case we would only need the cache to store the last value of the texture for each gobo, so it is effective when a non-texture parameter (like rotation) is changed. If the texture itself changes, there's little chance of it being in the cache anyway, and the user shouldn't mind a bit of a hit. I'd suggest a much lower limit - 100 say.

Reply:
I think storing gobo textures is a perfectly good use of 30GB of RAM! :P

No, what you suggest sounds good - one thing to keep in mind is that we allow not just texture reads but all kinds of osl manipulation. I've been playing with color grading the texture for example and caching the result for certain nodes being disabled seemed like a reasonable use of the cache to me too. The cache is definitely too large right now, though, and 100 seems like a good value. Sorry, I've been naively copying that number from the example you posted earlier without doing the maths.